staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go

```diff
@@ -151,11 +212,100 @@ func WithTypeAndObjectMeta(s *schema.Structural) *schema.Structural {
 	props["metadata"] = schema.Structural{
 		Generic: schema.Generic{Type: "object"},
 		Properties: map[string]schema.Structural{
-			"name": stringType,
+			"name":         stringType,
 			"generateName": stringType,
 		},
 	}
 	result.Properties = props
 
 	return result
-}
\ No newline at end of file
+}
+
+// estimateMinSizeJSON estimates the minimum size in bytes of the given schema when serialized in JSON.
+// minLength/minProperties/minItems is not currently taken into account, so the estimate is a little
+// smaller than the actual smallest size.
+func estimateMinSizeJSON(s *schema.Structural) int64 {
+	switch s.Type {
+	// take a comma into account for all cases since we'll generally call this function for arrays/maps
+	case "boolean":
+		// true,
+		return 5
+	case "number", "integer":
+		// 0,
+		return 2
+	case "string":
+		if s.ValueValidation != nil && s.ValueValidation.Format != "" && s.ValueValidation.Format != "byte" {
+			switch s.ValueValidation.Format {
+			case "duration":
+				return minDurationSizeJSON
+			case "date":
+				// add 1 to make room for a comma; since we're in estimateMinSizeJSON we need to account for
+				// this date being an array element
+				return dateSizeJSON + 1
+			case "date-time":
+				return minDatetimeSizeJSON
+			}
+		}
+		// "",
+		return 3
+	case "array":
+		// [],
+		return 3
+	case "object":
+		objSize := int64(3) // {},
+		// sum of all non-optional properties
+		if s.ValueValidation != nil {
+			for _, propName := range s.ValueValidation.Required {
+				prop := s.Properties[propName]
+				// add 3, 2 for quotations around the property name and 1 for the colon
+				objSize += int64(len(propName)) + estimateMinSizeJSON(&prop) + 3
+			}
+		}
+		return objSize
+	}
+	// TODO(DangerOnTheRanger): better error handling (We should never get here in normal operation)
+	return noMaxLength
+}
+
+// estimateMaxSizeJSON estimates the maximum number of elements that can fit in s considering request size
+// constraints.
+func estimateMaxSizeJSON(s *schema.Structural) int64 {
+	switch s.Type {
+	case "number", "integer":
+		// we're concerned with the theoretical largest number that could be written as a string given the
+		// request limit, not the largest 64-bit integer (which is significantly smaller), so this is the
+		// same case as a string without ValueValidation
+		return (maxRequestSizeBytes - 2)
```

Comment:
It's odd that we're saying that number/integer can be unbounded in size but that duration/date/date-time (which are just strings that happen to have a format assigned) are not.

There seems to be two possible approaches here:
- Don't assume any OpenAPI format validation is being enforced and treat all strings (regardless of format) as being unbounded in size as well. This doesn't reflect what's going to happen in CEL very accurately because something like a duration is converted to an int64 for CEL. But since this function is for JSON max size, it would be correct.
- Assume that OpenAPI format validation has been applied, in which case the largest numbers would be the float64 and int64 with the longest string representations.

I'm in favor of the later, but it has the cavaet that we should probably not run CEL if OpenAPIv3 validation is failing on the data sub-tree that the CEL expression is evaluating (at least for the type checking, required fields formats and max limit validation failures).
 
It's also odd that we have cases for number and integer here but that we don't set a maxlengh on them in SchemaDeclType.

Reply:
Yeah, I had been wondering about whether or not to handle numbers like the latter bullet point as well. I think it probably makes more sense to go with that route, so I'll go ahead and change it (I think some unit tests will need to have their expected results changed as well).

> It's also odd that we have cases for number and integer here but that we don't set a maxlengh on them in SchemaDeclType.

It does seem kind of odd, but I thought it made sense after mulling it over. I think both numbers and integers have implicit maximum sizes due to hardware restrictions, but that restriction needs to be explicitly taken into account during serialization, hence them being handled here.